keps/168-pending-workloads-visibility/README.md

```diff
@@ -0,0 +1,694 @@
+	// status of pending workloads in the local queue.
+	// +optional
+	PendingWorkloadsStatus *LocalQueuePendingWorkloadsStatus
+...
+}
+```
+
+### Cluster Queue API
+
+```golang
+// ClusterQueuePendingWorkload contains the information identifying a pending workload
+// in the cluster queue.
+type ClusterQueuePendingWorkload struct {
+	// Name indicates the name of the pending workload.
+	Name string
+
+	// Namespace indicates the name of the pending workload.
+	Namespace string
+}
+
+type ClusterQueuePendingWorkloadsStatus struct {
+	// Head contains the list of top pending workloads.
+	// +listType=map
+	// +listMapKey=name
+	// +listMapKey=namespace
+	// +optional
+	Head []ClusterQueuePendingWorkload
+
+	// LastChangeTime indicates the time of the last change of the structure.
+	LastChangeTime metav1.Time
+}
+
+// ClusterQueueStatus defines the observed state of ClusterQueueStatus
+type ClusterQueueStatus struct {
+...
+	// PendingWorkloadsStatus contains the information exposed about the current
+	// status of the pending workloads in the cluster queue.
+	// +optional
+	PendingWorkloadsStatus *ClusterQueuePendingWorkloadsStatus
+...
+}
+```
+
+### Configuration
+
+```golang
+// Configuration is the Schema for the kueueconfigurations API
+type Configuration struct {
+...
+	// QueueVisibility is configuration to expose the information about the top
+	// pending workloads.
+	QueueVisibility *QueueVisibility
+}
+
+type QueueVisibility struct {
+	// LocalQueues is configuration to expose the information
+	// about the top pending workloads in the local queue.
+	LocalQueues *LocalQueueVisibility
+
+	// ClusterQueues is configuration to expose the information
+	// about the top pending workloads in the cluster queue.
+	ClusterQueues *ClusterQueueVisibility
+
+	// UpdateInterval specifies the time interval for updates to the structure
+	// of the top pending workloads in the queues.
+	// Defaults to 5s.
+	UpdateInterval time.Duration
+}
+
+type LocalQueueVisibility struct {
+	// MaxCount indicates the maximal number of pending workloads exposed in the
+	// local queue status. The maximal value is 4000.
+	// Defaults to 10.
+	MaxCount int32
+}
+
+type ClusterQueueVisibility struct {
+	// MaxCount indicates the maximal number of pending workloads exposed in the
+	// cluster queue status. The maximal value is 4000.
+	// Defaults to 10.
+	MaxCount int32
+}
+```
+
+### In-memory snapshot of the ClusterQueue
+
+In order to be able to quickly compute the top pending workloads per LocalQueue,
+without a need for a prolonged read lock on the ClusterQueue, we create
+periodically in-memory snapshot of the ClusterQueue, organized as a map
+from the LocalQueue to the list of workloads belonging to the ClusterQueue,
+along with their positions. Then, the LocalQueue and ClusterQueue controllers
+do lookup into the cached structure.
+
+The snapshot is taken periodically in intervals configured in
```

Comment:
Can we sync snapshots in stages, not all at once?
In the case of many clusterqueues and localqueues already existing, I guess the kube-apiserver load will explode, and a prolonged read lock will occur immediately after starting or restarting the kueue-controller-manager.

Reply:
Good point, which was also raised earlier. I now propose to have a queue to snapshot-taking tasks. The queue will be processed by 5 workers. This should allow taking the snapshots in batches. Anyway, I think there is not so much risk as the read lock is only taken for the duration of copying the underlying heap data.

Reply:
right, you take the data, but sorting happens outside of the lock. That could be relatively fast if you use `copy`.

Reply:
SGTM